wallet-new/src/Cardano/Wallet/WalletLayer/Legacy.hs

```diff
@@ -1,43 +1,233 @@
+
 -- | Initialize the active wallet.
 -- The active wallet is allowed all.
 bracketActiveWallet
-    :: forall ctx m n a. (MonadMask m, WalletDbReader ctx n, MonadIO n, MonadThrow n)
-    => PassiveWalletLayer n
+    :: forall ctx m n a. (MonadMask n, MonadLegacyWallet ctx m)
+    => PassiveWalletLayer m
     -> WalletDiffusion
-    -> (ActiveWalletLayer n -> m a) -> m a
+    -> (ActiveWalletLayer m -> n a) -> n a
 bracketActiveWallet walletPassiveLayer walletDiffusion =
     bracket
       (return ActiveWalletLayer{..})
       (\_ -> return ())
 
+------------------------------------------------------------
+-- Wallet
+------------------------------------------------------------
+
+pwlCreateWallet
+    :: forall ctx m. (MonadLegacyWallet ctx m)
+    => NewWallet
+    -> m Wallet
+pwlCreateWallet NewWallet{..} = do
+
+    let spendingPassword = fromMaybe mempty $ coerce newwalSpendingPassword
+    let backupPhrase     = coerce newwalBackupPhrase
+
+    initMeta    <- CWalletMeta  <$> pure newwalName
+                                <*> migrate newwalAssuranceLevel
+                                <*> pure 0
+
+    let walletInit = CWalletInit initMeta backupPhrase
+
+    wallet      <- newWalletHandler newwalOperation spendingPassword walletInit
+    wId         <- migrate $ cwId wallet
+
+    -- Get wallet or throw if missing.
+    maybeThrow (MissingWallet wId) =<< pwlGetWallet wId
+  where
+    -- | We have two functions which are very similar.
+    newWalletHandler :: WalletOperation -> PassPhrase -> CWalletInit -> m CWallet
+    newWalletHandler CreateWallet  = newWallet
+    newWalletHandler RestoreWallet = restoreWallet
+
+
+pwlGetWalletIds
+    :: forall ctx m. (MonadLegacyWallet ctx m)
+    => m [WalletId]
+pwlGetWalletIds = do
+    ws          <- askWalletSnapshot
+    migrate $ getWalletAddresses ws
+
+pwlGetWallet
+    :: forall ctx m. (MonadLegacyWallet ctx m)
+    => WalletId
+    -> m (Maybe Wallet)
+pwlGetWallet wId = do
+    ws          <- askWalletSnapshot
+
+    cWId        <- migrate wId
+    wallet      <- V0.getWallet cWId
+
+    pure $ do
+        walletInfo  <- getWalletInfo cWId ws
+        migrate (wallet, walletInfo)
+
+pwlUpdateWallet
+    :: forall ctx m. (MonadLegacyWallet ctx m)
+    => WalletId
+    -> WalletUpdate
+    -> m Wallet
+pwlUpdateWallet wId wUpdate = do
+    walletDB    <- askWalletDB
+
+    cWId        <- migrate wId
+    cWMeta      <- migrate wUpdate
+
+    -- Update the data
+    setWalletMeta walletDB cWId cWMeta
+
+    -- Get wallet or throw if missing.
+    maybeThrow (MissingWallet wId) =<< pwlGetWallet wId
+
+-- | Seems silly, but we do need some sort of feedback from
+-- the DB.
+pwlDeleteWallet
+    :: forall ctx m. (MonadLegacyWallet ctx m)
+    => WalletId
+    -> m Bool
+pwlDeleteWallet wId = do
+    cWId        <- migrate wId
+    -- TODO(ks): It would be better to catch specific @Exception@.
+    -- Maybe @try@?
+    catchAll (const True <$> V0.deleteWallet cWId) (const . pure $ False)
```

Comment:
Does the DB throw upon deleting an non-existing wallet? I don't see why we would swallow db exceptions here.

Reply:
Let's definitely *not* swallow exceptions here, without at least reporting on them. 

I have seen the `V0` code throw a `WalletError` type defined in `wallet/src/Pos/Wallet/Web/Errors/Types.hs` and not much else.

Reply:
Good catch guys, already fixed in https://github.com/input-output-hk/cardano-sl/pull/2778 (the PR following this one).

Reply:
Agree, we should fail appropriately, if not with an exception, with a proper `Either` type to be reflected in the type signature of `pwlDeleteWallet`.